Keep covering clauses of function definitions

jespercockx opened the pull request:
Currently, we only keep the covering clauses for checking IApply confluence, but they would also be useful to keep for use by backends (e.g. the Dedukti backend that @GuillaumeGen is working on). This PR changes Agda so it keeps the covering clauses and stores them in the interface for later use.

Can be merged after Travis succeeds.

jespercockx committed:
[ refactor ] Always keep covering clauses + store in interface

jespercockx commented:
Does anyone know why the Travis build is timing out on GHC 8.0 (and only 8.0)?

jespercockx commented:
When I run it on my own PC (using GHC 8.0.2), agda runs out of heap:
```
  Checking Data.Vec.Relation.Unary.Any.Properties (/home/jesper/agda/std-lib/src/Data/Vec/Relation/Unary/Any/Properties.agda).
agda: Heap exhausted;
agda: Current maximum heap size is 1610612736 bytes (1536 MB).
agda: Use `+RTS -M<size>' to increase it.
 458,938,669,560 bytes allocated in the heap
  54,521,359,992 bytes copied during GC
   1,585,255,328 bytes maximum residency (408 sample(s))
      11,855,552 bytes maximum slop
            1576 MB total memory in use (8 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     399526 colls,     0 par   85.802s  85.961s     0.0002s    0.1132s
  Gen  1       408 colls,     0 par   1930.236s  1931.733s     4.7346s    10.2548s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.000s  (  0.000s elapsed)
  MUT     time  276.135s  (276.948s elapsed)
  GC      time  2016.038s  (2017.694s elapsed)
  EXIT    time    0.000s  (  0.000s elapsed)
  Total   time  2292.175s  (2294.643s elapsed)

  Alloc rate    1,662,008,079 bytes per MUT second

  Productivity  12.0% of total user, 12.1% of total elapsed

gc_alloc_block_sync: 0
whitehole_spin: 0
gen[0].sync: 0
gen[1].sync: 0

real	38m14,760s
user	38m6,996s
sys	0m5,286s
Makefile:314: recipe for target 'library-test' failed
make: *** [library-test] Error 251
```
Since the covering clauses are no longer erased from the signature, could this lead to some kind of memory leak and cause the heap exhaustion?

jespercockx commented:
I'm trying again to run on Travis with 2GB heap for the standard library (instead of 1.5GB), just to see if it makes a difference.

gallais commented:
FYI we are using 2.5G of heap on the stdlib travis instance:
https://github.com/agda/agda-stdlib/blob/master/.travis.yml#L67

jespercockx commented:
It fixes the problem on my own PC but not on Travis: the build simply times out on       
```
      with-stdlib
        AllStdLib:                   
```

jespercockx commented:
I'm not sure what to do about this: apparently keeping the extra information of the covering clauses in the interface makes Agda significantly slower (at least of the `AllStdLib` compiler test). On the other hand, this is useful information for backend developers that is not easy to reconstruct afterwards. Two options that could work:

1. Add a flag to keep the covering clauses in the interface. This would require re-checking the whole library before getting access to this information, which is annoying.
2. Add a function to re-run the coverage check and return the covering clauses. Not quite as expensive as 1. because we only re-run the coverage check, but this means the coverage check is run twice *even when we know that we will need the covering clauses the first time*.

I suppose another option would be to combine 1. and 2. by keeping the covering clauses when the flag is enabled and reconstructing it otherwise. But this seems to be a too complex solution for a simple problem.

andreasabel commented:
@Saizan : Can you confirm it is save to remove the `Closure`? 

@jespercockx : Do you have some data on how the size of the .agdai files increases?  Maybe it is not so bad because the introduction of sharing during serialization.  
Also, how do the timings for the standard-library test change?

Saizan commented:
@andreasabel @jespercockx The main point of using a ```Closure``` was to keep the ```scCheckpoints :: Map CheckpointId Substitution``` around, as is done for ```inferMissingClause```. The ```Closure``` was capturing the content of that field.

However I'm not an expert in checkpoints, I just remember I had failures in the testsuite when running conversion checks in ```checkIApplyConfluece``` without the closure. Maybe the changes to ```Defn``` instances in the same commit took care of that?

jespercockx commented:
Yes, the point of these other changes was to make it unnecessary to use a closure by treating the covering clauses in the same way as the regular clauses. I tested it with the cubical library and everything went through fine, so I assume I didn't mess up anything too bad.

jespercockx commented:
The difference in size of the `.agdai` files is not that big: I just tested and it went up from 65MB to 67MB (details here: https://gist.github.com/jespercockx/c28a4792b6d3619c167e22198d73b07c).

Also note that the timeout doesn't happen while checking the standard library itself, but rather during the `AllStdLib` of the Compiler Tests. So it seems the problem lies somewhere else.

jespercockx commented:
Rebased this onto master to see if the memory issues I had in 2019 have disappeared.

jespercockx commented:
I have updated this PR again to the current master. Either the performance problem was resolved, or we have dropped support for the version of GHC where this was a problem. Is anyone opposed to merging this in?

andreasabel commented:
AFAICT, we haven't dropped any GHC version since, but we have dropped Travis CI.

But note that we are only running the testsuite now under the latest GHC.  Previously, for Travis, there was the `[complete tests]` tag to trigger running the tests on all GHC versions.

jespercockx commented:
It seems you're right, I've compiled Agda again with GHC 8.0.2 and this PR, and now checking the standard library gets stuck on `Checking Algebra.Morphism.Construct.Composition (/home/jesper/agda/std-lib/src/Algebra/Morphism/Construct/Composition.agda).` So we'll have to find another solution.

andreasabel commented:
Following our current policy, we would drop GHC 8.0 only after April 2023 (EOL for Ubuntu 18.04 LTS).

jespercockx committed:
[ re #3909 ] Add flag --keep-covering-clauses to toggle discarding of covering clauses

jespercockx commented:
Since I'd like to have this in the release of Agda 2.6.3, I have hidden the option under a flag for now. Once we drop support for GHC 8.0, we could consider enabling this flag by default if desired.

andreasabel commented:
Alternatively, you could have a __cabal flag__ for switching this functionality and switch this off in the CI for GHC 8.0.2 and on otherwise.  Default could be _on_ as most will build Agda with a more recent GHC.
